### DISCUSSION TITLE
List accessor syntax

### DISCUSSION THREAD

**Discussion by siddhantk232** [fastn-stack/fastn#1985] [Ideas & RFCs]
I don't think we have list accessor syntax to get an item of a list by it's index. And How do we create a list of integers or strings?

```ftd

-- integer list user-choice-of-ids:

;; what goes in here to create a list item of type `integer`?

-- end: user-choice-of-ids

-- integer default-id: $user-choice-of-ids[0]

```

**Comment by amitu** [comment]
```ftd
-- integer list user-choice-of-ids:
```

I am not happy with this any more and prefer:

```ftd
-- list<integer> user-choice-of-ids:
```

Now. This is so we can add more complex types like result, [remote-data](https://package.elm-lang.org/packages/krisajenkins/remotedata/latest/RemoteData), dicts etc. And also so functions / records etc can be take generic input / output. 


```
-- list<integer> user-choice-of-ids:

-- integer: 1
-- integer: 2

-- end: user-choice-of-ids
```

And using the `it` keyword:

```
-- list<integer> user-choice-of-ids:

-- it: 1
-- it: 2

-- end: user-choice-of-ids
```

```
-- integer default-id: user-choice-of-ids[0]
```

For integer we will not need `$` to reference variable going forward. Only string needs special handling for variable names, and for string we want string template / format string like syntax for variable lookup (also string calls a `.to_string()` on each dynamic thing). Also string will keep tree structure, so when a variable embedded in a long text changes, we do not have to re-render entire long text, but only the text fragment where the variable's to_string is embedded.

**Reply by amitu** [reply]
```ftd
-- list<integer> user-choice-of-ids: { [1, 2] }
```

Should also work. We went from `p-script` to `f-script` using the `{}`, and in `f-script` you create an integer list using `[1, 2]` syntax.

**Reply by amitu** [reply]
Actually since we want to allow: 

```ftd
-- integer x: 1 + 2
```

Using the rule "`r-value` of non-string `l-variable` is implicitly `f-script`", we should also allow

```ftd
-- list<string> user-choice-of-ids: [1, 2]
```

Problem will be for `list<string>`, where you continue to use the 

```ftd
-- list<integer> sids:

-- it: 1
-- it: 2
-- it:

this is a list of strings after all

-- end: sids
```
